### PR TITLE
layers: Add VK_QCOM_tile_shading dispatch VUIDs (1/2)

### PR DESCRIPTION
**Pull request by ZZKQcom**
# Introduction
VK_QCOM_tile_shading dispatch/draw VUIDs and unit tests. Part I. 6 VUIDs totally.

# VUID coverage
| Test | VUID |
|------|------|
| `DispatchTileButPerTileExecutionModelNotEnabled` | `VUID-vkCmdDispatchTileQCOM-None-10668` |
| `DispatchTileButPerTileDispatchFeatureNotEnabled` | `VUID-vkCmdDispatchTileQCOM-None-10669` |
| `TileShadingDrawButHasActiveGeometryShader` | `VUID-vkCmdDraw-None-10678` |
| `TileShadingDrawButHasActiveGeometryShaderObject` | `VUID-vkCmdDraw-None-10678` |
| `PerTileDrawButAccessImageWithFeedbackLoopLayout` | `VUID-vkCmdDraw-None-10679` |
| `ImageDescriptorMismatchTileMemory` | `VUID-vkCmdDispatchTileQCOM-commandBuffer-10746` |
| `DispatchInsidePerTileExecutionModelButPerTileDispatchFeatureNotEnabled` | `VUID-vkCmdDispatch*-None-10674` |
| `DrawInsidePerTileExecutionModelButPerTileDrawFeatureNotEnabled` | `VUID-vkCmdDraw*-None-10677` |
| `DispatchInsidePerTileExecutionModelAndTileShadingRenderPassScopes` | Positive unit test |

# Note
- `vkCmdDispatchTileQCOM` shares the same VUIDs as `vkCmdDispatch`, expect `VUID-vkCmdDispatch-groupCountX-00386` and `VUID-vkCmdDispatch-groupCountY-00387` and `VUID-vkCmdDispatch-groupCountZ-00388`.
- `vkCmdDispatchTileQCOM` has additional VUIDs `VUID-vkCmdDispatchTileQCOM-None-10668` and `VUID-vkCmdDispatchTileQCOM-None-10669`.
- The positive unit tests for tile shading draw and dispatch will be completed in dispatch/draw part 2.

### CODE REVIEW

**Comment by spencer-lunarg** [comment]
> The negative unit tests for VUID-vkCmdDispatch*-None-10674 and VUID-vkCmdDraw*-None-10677 were not added because these errors are caught earlier by the stateless validation check for VUID-vkCmdBeginPerTileExecutionQCOM-None-10665 in manual_PreCallValidateCmdBeginPerTileExecutionQCOM, causing the test to exit before reaching those VUIDs.

@ZZKQcom are you saying there is no way to actually hit these 2 VUs? If so, I want to know, because we have removed VUs in the past. Can you provide which VUs are you hitting first?

File: layers/core_checks/cc_render_pass.cpp [hunk 1]
```diff
@@ -4625,6 +4625,12 @@ bool CoreChecks::InsideRenderPass(const vvl::CommandBuffer& cb_state, const Loca
         return skip;
     }
 
+    if (cb_state.per_tile_execution_model_enabled &&
+        (loc.function == vvl::Func::vkCmdDispatch || loc.function == vvl::Func::vkCmdDispatchIndirect ||
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_render_pass.cpp]: 1. Add a comment why we are skipping here
2. Use `IsCommandDispatch`

File: layers/core_checks/cc_drawdispatch.cpp [hunk 2]
```diff
@@ -1651,9 +1669,43 @@ bool CoreChecks::ValidateActionState(const LastBound& last_bound_state, const Dr
...
         skip |= InsideRenderPass(cb_state, loc);
 
+        if (cb_state.per_tile_execution_model_enabled && !enabled_features.tileShadingPerTileDispatch) {
+            skip |= LogError(CreateActionVuid(loc.function, vvl::ActionVUID::PER_TILE_DISPATCH_ENABLED_10674),
+                             cb_state.GetObjectList(bind_point), loc,
+                             "the per-tile execution model is enabled in this command buffer, but "
+                             "VkPhysicalDeviceTileShadingFeaturesQCOM::tileShadingPerTileDispatch feature isn't enabled.");
+        }
+
+        if (pipeline && cb_state.active_render_pass && cb_state.active_render_pass->has_tile_shading_enabled) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_drawdispatch.cpp]: no reason to have `10678` done twice

make a new function like `ValidateActionStateTileShading` that is called once and anything you added to this `ValidateActionState` should go there

Comment (ZZKQcom) [reply layers/core_checks/cc_drawdispatch.cpp]: Move to `ValidateActionStateTileShading` and call it in `ValidateActionState`.

File: layers/core_checks/cc_render_pass.cpp (hunk 1 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_render_pass.cpp]: Refactor the implementation.
I only found `VUID-vkCmdDispatch*-None-10672` in `vkCmdDispatch`, `vkCmdDispatchIndirect`, `vkCmdDispatchBase` and `vkCmdDispatchTileQCOM` commands, but seems `vvl::IsCommandDispatch` includes additional `vkCmdDispatchGraphAMDX`, `vkCmdDispatchGraphIndirectAMDX`, `vkCmdDispatchGraphIndirectCountAMDX` and `vkCmdDispatchDataGraphARM`.

**Comment by ZZKQcom** [comment]
Hi, @spencer-lunarg 
For the missing unit tests of `VUID-vkCmdDispatch*-None-10674` and `VUID-vkCmdDraw*-None-10677`, I actually made a mistake.

In my original unit tests, both `tileShadingPerTileDispatch` and `tileShadingPerTileDraw` were disabled, so `vkCmdBeginPerTileExecutionQCOM` hit `VUID-vkCmdBeginPerTileExecutionQCOM-None-10665` firstly and the process exited and the command buffer never entered per-tile execution scope.

These VUIDs are reachable, but the setup needs to enable the opposite feature:
- for `VUID-vkCmdDispatch*-None-10674`, keep `tileShadingPerTileDispatch` disabled but enable `tileShadingPerTileDraw`
- for `VUID-vkCmdDraw*-None-10677`, keep `tileShadingPerTileDraw` disabled but enable `tileShadingPerTileDispatch`

I have added:
- `NegativeTileShading.DispatchInsidePerTileExecutionModelButPerTileDispatchFeatureNotEnabled`
- `NegativeTileShading.DrawInsidePerTileExecutionModelButPerTileDrawFeatureNotEnabled`
in the new commit and the tests pass.

**Comment by ZZKQcom** [comment]
I have updated the implementation based on your suggestions, added negative unit tests for `VUID-10674` and `VUID-10677`, and squashed the two commits into one.

**Comment by spencer-lunarg** [comment]
@ZZKQcom thanks, its SDK week, trying to squeeze things in, will take a look again Monday! Sorry for the delay

File: layers/core_checks/cc_drawdispatch.cpp [hunk 3]
```diff
@@ -1656,6 +1673,51 @@ bool CoreChecks::ValidateActionState(const LastBound& last_bound_state, const Lo
...
+    return skip;
+}
+
+bool CoreChecks::ValidateActionStateTileShading(const LastBound& last_bound_state, const Location& loc) const {
+    bool skip = false;
+    const vvl::CommandBuffer& cb_state = last_bound_state.cb_state;
+    const vvl::Pipeline* pipeline = last_bound_state.pipeline_state;
+    const VkPipelineBindPoint bind_point = last_bound_state.bind_point;
+
+    if (bind_point == VK_PIPELINE_BIND_POINT_GRAPHICS) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_drawdispatch.cpp]:
```c++
    if (cb_state.per_tile_execution_model_enabled && !enabled_features.tileShadingPerTileDispatch) {
        if (bind_point == VK_PIPELINE_BIND_POINT_GRAPHICS || bind_point == VK_PIPELINE_BIND_POINT_COMPUTE) {
            vvl::ActionVUID action_vuid = bind_point == VK_PIPELINE_BIND_POINT_GRAPHICS ? vvl::ActionVUID::PER_TILE_DRAW_ENABLED_10677 : vvl::ActionVUID::PER_TILE_DISPATCH_ENABLED_10674;
            skip |= LogError(CreateActionVuid(loc.function, action_vuid),
                             cb_state.GetObjectList(bind_point), loc,
                             "the per-tile execution model is enabled in this command buffer, but "
                             "VkPhysicalDeviceTileShadingFeaturesQCOM::tileShadingPerTileDraw feature isn't enabled.");
        }
    }
```

File: layers/core_checks/cc_drawdispatch.cpp [hunk 4]
```diff
@@ -1656,6 +1673,51 @@ bool CoreChecks::ValidateActionState(const LastBound& last_bound_state, const Lo
...
+    } else if (bind_point == VK_PIPELINE_BIND_POINT_COMPUTE) {
+        if (cb_state.per_tile_execution_model_enabled && !enabled_features.tileShadingPerTileDispatch) {
+            skip |= LogError(CreateActionVuid(loc.function, vvl::ActionVUID::PER_TILE_DISPATCH_ENABLED_10674),
+                             cb_state.GetObjectList(bind_point), loc,
+                             "the per-tile execution model is enabled in this command buffer, but "
+                             "VkPhysicalDeviceTileShadingFeaturesQCOM::tileShadingPerTileDispatch feature isn't enabled.");
+        }
+    }
+
+    if (pipeline && cb_state.active_render_pass && cb_state.active_render_pass->has_tile_shading_enabled) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_drawdispatch.cpp]: leave a comment why there is no ShaderObject path here.

```c
// VK_QCOM_tile_shading doesn't support shader object
```

is a valid comment, if that is true

File: layers/core_checks/cc_drawdispatch.cpp [hunk 5]
```diff
@@ -1656,6 +1673,51 @@ bool CoreChecks::ValidateActionState(const LastBound& last_bound_state, const Lo
...
+        VkShaderStageFlags allowed_shader_stages = 0;
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_drawdispatch.cpp]:
```suggestion
        const VkShaderStageFlags allowed_shader_stages = VK_SHADER_STAGE_VERTEX_BIT | VK_SHADER_STAGE_FRAGMENT_BIT | VK_SHADER_STAGE_COMPUTE_BIT;
```

You don't need to worry about `pipeline->active_shaders` having the "wrong" type of shader stages for the bind point

File: layers/core_checks/cc_drawdispatch.cpp [hunk 6]
```diff
@@ -1656,6 +1673,51 @@ bool CoreChecks::ValidateActionState(const LastBound& last_bound_state, const Lo
...
+        if (bind_point == VK_PIPELINE_BIND_POINT_GRAPHICS) {
+            allowed_shader_stages = VK_SHADER_STAGE_VERTEX_BIT | VK_SHADER_STAGE_FRAGMENT_BIT;
+        } else if (bind_point == VK_PIPELINE_BIND_POINT_COMPUTE) {
+            allowed_shader_stages = VK_SHADER_STAGE_COMPUTE_BIT;
+        }
+
+        if (allowed_shader_stages != 0 && (pipeline->active_shaders & ~allowed_shader_stages) != 0) {
+            skip |= LogError(CreateActionVuid(loc.function, vvl::ActionVUID::TILE_SHADING_SHADER_STAGE_10678),
+                             cb_state.GetObjectList(bind_point), loc,
+                             "shader stages (%s) include stages not allowed for the (%s) bind point "
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_drawdispatch.cpp]:
```suggestion
                             "shader stages (%s) include stages not allowed "
```

it is obvious that Computer shaders are in compute bind points

File: layers/core_checks/cc_render_pass.cpp [hunk 7]
```diff
@@ -4631,6 +4632,13 @@ bool CoreChecks::InsideRenderPass(const vvl::CommandBuffer& cb_state, const Loca
         return skip;
     }
 
+    // Note: According to the spec, inside the per-tile execution model scope, the following dispatch commands
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_render_pass.cpp:4635]: 1. Great comment!
2. Please provide a positive test doing a simple `vkCmdDispatch` inside a render pass with tile shading enabled to help make sure if this was removed, the positive test would fail

File: layers/core_checks/cc_drawdispatch.cpp (hunk 3 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_drawdispatch.cpp]: Thanks for the suggestion! However, I think merging the two conditions may introduce a bug here.

`tileShadingPerTileDraw` and `tileShadingPerTileDispatch` are two independent feature bits, the graphics bind point checks `tileShadingPerTileDraw` (`VUID 10677`), while the compute bind point checks `tileShadingPerTileDispatch` (`VUID 10674`).

File: layers/core_checks/cc_drawdispatch.cpp (hunk 4 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_drawdispatch.cpp]: The `VK_QCOM_tile_shading` spec does not explicitly say whether `shader objects` are supported. I think the `VUID-10678` check should therefore account for them as well. I also added a corresponding negative unit test: `TileShadingDrawButHasActiveGeometryShaderObject`.

File: layers/core_checks/cc_drawdispatch.cpp (hunk 5 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_drawdispatch.cpp]: Fix according to the suggestion.

File: layers/core_checks/cc_drawdispatch.cpp (hunk 6 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_drawdispatch.cpp]: Fix according to the suggestion.

File: layers/core_checks/cc_render_pass.cpp (hunk 7 above)
Comment (ZZKQcom) [reply layers/core_checks/cc_render_pass.cpp:4635]: Add positive unit test `DispatchInsidePerTileExecutionModelAndTileShadingRenderPassScopes`.

File: layers/core_checks/cc_drawdispatch.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_drawdispatch.cpp]: yep, didn't realize the `tileShadingPerTileDraw` and `tileShadingPerTileDispatch` were not the same word

**Review by spencer-lunarg — APPROVED** [review]
